File: Site 3/GPT4/codigo_html.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.NoSuchElementException;
import java.time.Duration;
import java.util.List;

public class SwagLabsLoginTest {
    private static final String BASE_URL = "https://www.saucedemo.com/v1/";
    private static final String VALID_PASSWORD = "secret_sauce";
    private static final String[] VALID_USERNAMES = {"standard_user", "locked_out_user", "problem_user", "performance_glitch_user"};
    
    public static void main(String[] args) {
        // Set up Chrome WebDriver
        System.setProperty("webdriver.chrome.driver", "path_to_chromedriver");
        WebDriver driver = new ChromeDriver();
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));

        try {
            // Verify page structure
            verifyPageStructure(driver);
            
            // Test successful login
            testSuccessfulLogin(driver);
            
            // Test login failure
            testLoginFailure(driver);
            
            // Test error message accuracy
            testErrorMessages(driver);
        } finally {
            // Close the browser
            driver.quit();
        }
    }

    private static void verifyPageStructure(WebDriver driver) {
        driver.get(BASE_URL);

        // Check for the username and password fields
        assertElementPresent(driver, By.id("user-name"), "Username field");
        assertElementPresent(driver, By.id("password"), "Password field");

        // Check for the login button
        assertElementPresent(driver, By.id("login-button"), "Login button");

        // Check for "Accepted usernames are:" and "Password for all users:"
        assertElementPresent(driver, By.xpath("//h4[text()='Accepted usernames are:']"), "Accepted usernames label");
        assertElementPresent(driver, By.xpath("//h4[text()='Password for all users:']"), "Password label");

        // Check for the four usernames and one password
        List<WebElement> usernameElements = driver.findElements(By.xpath("//div[@id='login_credentials']/text()"));
        if (usernameElements.size() != 4) {
            throw new AssertionError("Expected 4 accepted usernames, found " + usernameElements.size());
        }
        WebElement passwordElement = driver.findElement(By.xpath("//div[@class='login_password']"));
        if (!passwordElement.getText().contains(VALID_PASSWORD)) {
            throw new AssertionError("Expected password not found.");
        }
    }

    private static void testSuccessfulLogin(WebDriver driver) {
        for (String username : VALID_USERNAMES) {
            driver.get(BASE_URL);
            performLogin(driver, username, VALID_PASSWORD);
            
            // Assert successful login - example placeholder assertion
            // Assuming redirection or message, adjust the locator accordingly.
            assertElementPresent(driver, By.className("success-message"), "Success message");

            driver.navigate().refresh();
        }
    }

    private static void testLoginFailure(WebDriver driver) {
        driver.get(BASE_URL);
        performLogin(driver, "invalid_user", VALID_PASSWORD);
        assertElementPresent(driver, By.className("error-message"), "Error message for invalid username");
        
        performLogin(driver, VALID_USERNAMES[0], "wrong_password");
        assertElementPresent(driver, By.className("error-message"), "Error message for invalid password");
    }

    private static void testErrorMessages(WebDriver driver) {
        driver.get(BASE_URL);
        performLogin(driver, "invalid_user", "wrong_password");
        String errorMessage = driver.findElement(By.className("error-message"))
                                    .getText();
        if (!errorMessage.equals("Invalid username or password.")) {
            throw new AssertionError("Unexpected error message: " + errorMessage);
        }
    }

    private static void performLogin(WebDriver driver, String username, String password) {
        WebElement usernameField = driver.findElement(By.id("user-name"));
        WebElement passwordField = driver.findElement(By.id("password"));
        WebElement loginButton = driver.findElement(By.id("login-button"));

        usernameField.clear();
        usernameField.sendKeys(username);
        passwordField.clear();
        passwordField.sendKeys(password);
        loginButton.click();
    }

    private static void assertElementPresent(WebDriver driver, By locator, String elementName) {
        try {
            WebElement element = driver.findElement(locator);
            if (!element.isDisplayed()) {
                throw new AssertionError(elementName + " is not visible.");
            }
        } catch (NoSuchElementException e) {
            throw new AssertionError(elementName + " is not present.");
        }
    }
}



/*
Key Features of the Code:
Page Structure Verification: Ensures all required elements (username, password fields, and labels) are present.
Successful Login Testing: Loops through all valid usernames and verifies successful login.
Login Failure Handling: Tests with invalid credentials to check for error messages.
Error Message Verification: Ensures correct error messages for invalid inputs.
Assertions: Validates expected outcomes using assertions for each test scenario.
Make sure to update "path_to_chromedriver" with the actual path to your ChromeDriver executable.

*/

File: Site 2/GPT4/codigo_html.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

# Initialize the WebDriver
driver = webdriver.Chrome()

# Define the application URL
base_url = "https://wavingtest.github.io/system-healing-test/"

try:
    # Open the login page
    driver.get(base_url)

    # Implicit wait for dynamic elements
    driver.implicitly_wait(5)

    # Test: Verify page structure and content
    assert "Login" in driver.title, "Page title does not match."
    assert driver.find_element(By.CLASS_NAME, "banner"), "Banner not found."
    assert driver.find_element(By.ID, "username-555"), "Username field not found."
    assert driver.find_element(By.ID, "password-88888"), "Password field not found."
    assert driver.find_element(By.ID, "btnLogin65"), "Login button not found."
    assert driver.find_element(By.LINK_TEXT, "Esqueceu a sua senha?"), "Forgot Password link not found."
    assert driver.find_element(By.LINK_TEXT, "Criar uma conta"), "Create Account link not found."
    print("Page structure verified successfully.")

    # Test: Login functionality
    username_field = driver.find_element(By.ID, "username-555")
    password_field = driver.find_element(By.ID, "password-88888")
    login_button = driver.find_element(By.ID, "btnLogin65")
    error_message = driver.find_element(By.ID, "errorMessage5")

    # Successful login
    username_field.clear()
    password_field.clear()
    username_field.send_keys("validUser")
    password_field.send_keys("validPassword")
    login_button.click()
    # Wait for success toast to appear
    WebDriverWait(driver, 5).until(
        EC.visibility_of_element_located((By.ID, "toast"))
    )
    print("Successful login test passed.")

    # Failed login
    username_field.clear()
    password_field.clear()
    username_field.send_keys("invalidUser")
    password_field.send_keys("invalidPassword")
    login_button.click()
    # Verify error message
    WebDriverWait(driver, 5).until(
        lambda d: error_message.is_displayed() and "error" in error_message.text.lower()
    )
    print("Failed login test passed.")

    # Test: Forgot Password hyperlink redirection
    forgot_password_link = driver.find_element(By.LINK_TEXT, "Esqueceu a sua senha?")
    forgot_password_link.click()
    WebDriverWait(driver, 5).until(
        EC.url_contains("password.html")
    )
    assert "password.html" in driver.current_url, "Forgot Password redirection failed."
    print("Forgot Password link test passed.")

    # Navigate back
    driver.back()

    # Test: Create Account hyperlink redirection
    create_account_link = driver.find_element(By.LINK_TEXT, "Criar uma conta")
    create_account_link.click()
    WebDriverWait(driver, 5).until(
        EC.url_contains("account.html")
    )
    assert "account.html" in driver.current_url, "Create Account redirection failed."
    print("Create Account link test passed.")

    # Navigate back
    driver.back()

    # Test: Social media icon redirections
    social_icons = driver.find_elements(By.CSS_SELECTOR, ".social img")
    social_links = ["facebook.com", "google.com", "twitter.com", "github.com"]

    for icon, link in zip(social_icons, social_links):
        icon.click()
        WebDriverWait(driver, 5).until(EC.new_window_is_opened)
        driver.switch_to.window(driver.window_handles[-1])  # Switch to the new tab
        assert link in driver.current_url, f"Redirection to {link} failed."
        driver.close()
        driver.switch_to.window(driver.window_handles[0])  # Switch back to main tab
    print("Social media link tests passed.")

finally:
    # Close the browser
    driver.quit()


"""
Key Features
Structure Verification: Validates the presence of key elements on the page.
Login Tests:
Validates successful login through the presence of the success toast.
Validates failed login by checking for the error message.
Hyperlink Redirections:
Ensures the Forgot Password and Create Account links redirect to their respective pages.
Social Media Links:
Verifies each icon redirects to the correct platform.
Explicit Waits: Handles dynamic content loading.
Assertions: Validates expected outcomes at each step.

"""